Client/Frontend/Sync/SyncSettingsTableViewController.swift

```diff
@@ -39,85 +36,141 @@ class SyncSettingsTableViewController: UITableViewController {
         tableView.tableHeaderView?.sizeToFit()
         
-        if disableBackButton {
+        if showDoneButton {
             navigationController?.interactivePopGestureRecognizer?.isEnabled = false
-            
             navigationItem.setHidesBackButton(true, animated: false)
             navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self,
                                                                 action: #selector(doneTapped))
         }
     }
     
-    // MARK: - Button actions.
+    // MARK: Private
     
-    @objc func doneTapped() {
+    private struct BraveSyncDevice: Codable {
+        let chromeVersion: String
+        let hasSharingInfo: Bool
+        let id: String
+        let guid: String
+        let isCurrentDevice: Bool
+        let supportsSelfDelete: Bool
+        let lastUpdatedTimestamp: TimeInterval
+        let name: String?
+        let os: String
+        let sendTabToSelfReceivingEnabled: Bool
+        let type: String
+        
+        func remove() {
+            // Devices other than `isCurrentDevice` can only be deleted if
+            // `supportsSelfDelete` is true. Attempting to delete another device
+            // from the sync chain that does not support it, will crash or
+            // have undefined behaviour as the other device won't know that
+            // it was deleted.
+            BraveSyncAPI.shared.removeDeviceFromSyncGroup(deviceGuid: self.guid)
+        }
+    }
+    
+    private var syncDeviceObserver: AnyObject?
+    private var devices = [BraveSyncDevice]()
+    
+    private enum Sections: Int {
+        case deviceList, deviceActions, syncTypes
+    }
+    
+    private enum SyncDataTypes: Int {
+        case bookmarks = 0, history
+    }
+    
+    /// A different logic and UI is used depending on what device was selected to remove and if it is the only device
+    /// left in the Sync Chain.
+    private enum DeviceRemovalType {
+        case lastDeviceLeft, currentDevice, otherDevice
+    }
+    
+    /// After synchronization is completed, user needs to tap on `Done` to go back.
+    /// Standard navigation is disabled then.
+    private var showDoneButton = false
+    
+    // MARK: Actions
+    
+    @objc private func doneTapped() {
         navigationController?.popToRootViewController(animated: true)
     }
     
-    private func addAnotherDeviceAction() {
-        let view = SyncSelectDeviceTypeViewController()
-        view.syncInitHandler = { title, type in
-            let view = SyncAddDeviceViewController(title: title, type: type)
-            view.doneHandler = {
-                self.navigationController?.popToViewController(self, animated: true)
+    private func presentAlertPopup(for type: DeviceRemovalType, device: BraveSyncDevice) {
+        var title: String?
+        var message: String?
+        var removeButtonName: String?
+        let deviceName = device.name ?? Strings.syncRemoveDeviceDefaultName
```

Comment:
@soner-yuksel - do you know if the `device-name` is sanitized before presenting in UI? I've come across cases where `<script>` tags in device name causes script execution.

Reply:
This was an existing method which was moved in the file from previous position.
However the device_name is not a user entry field, it is fetched using `BraveSyncAPI` and underneath this information is fetched from `DeviceInfoSyncServiceFactory`

Reply:
I verified that the `device-name` is not sanitized. I wasn't able to cause an injection, but as a defense in depth strategy what do you think about encoding the string before printing out the device name.